src/uct/ib/rc/verbs/rc_verbs_ep.c

```diff
@@ -550,6 +550,111 @@ ucs_status_t uct_rc_verbs_ep_fc_ctrl(uct_ep_t *tl_ep, unsigned op,
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        status = uct_rc_iface_qp_init(&iface->super, ep->tm_qp);
+        if (status != UCS_OK) {
+            ret = ibv_destroy_qp(ep->tm_qp);
+            if (ret) {
+                ucs_warn("ibv_destroy_qp() returned %d: %m", ret);
+            }
+            return status;
+        }
+        uct_rc_iface_add_ep(&iface->super, &ep->super, ep->tm_qp->qp_num);
+    }
+#endif
+    return UCS_OK;
+}
+
+static void uct_rc_verbs_ep_tag_qp_destroy(uct_rc_verbs_ep_t *ep)
+{
+#if HAVE_IBV_EX_HW_TM
+    uct_rc_verbs_iface_t *iface = ucs_derived_of(ep->super.super.super.iface,
+                                                 uct_rc_verbs_iface_t);
+    if (UCT_RC_VERBS_TM_ENABLED(iface)) {
+        if (ibv_destroy_qp(ep->tm_qp)) {
+            ucs_warn("failed to destroy TM RNDV QP: %m");
+        }
+        uct_rc_iface_remove_ep(&iface->super, ep->tm_qp->qp_num);
+    }
+#endif
+}
+
+#if HAVE_IBV_EX_HW_TM
+ucs_status_t uct_rc_verbs_ep_tag_get_address(uct_ep_h tl_ep,
+                                             uct_ep_addr_t *addr)
+{
+    ucs_status_t status;
+    uct_rc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    uct_rc_verbs_ep_tm_address_t *tm_addr = (uct_rc_verbs_ep_tm_address_t*)addr;
+
+    status = uct_rc_ep_get_address(tl_ep, (uct_ep_addr_t*)&tm_addr->super);
+    if (status == UCS_OK) {
+        uct_ib_pack_uint24(tm_addr->tm_qp_num, ep->tm_qp->qp_num);
+        /* Redefine address type */
+        tm_addr->super.type = UCT_RC_EP_ADDR_TYPE_TM;
+    }
+    return status;
+}
+
+/* For HW TM we need 2 QPs, one of which will be used by the device for
+ * RNDV offload (for issuing RDMA reads and sending RNDV ACK). No WQEs should
+ * be posted to the send side of the QP which is owned by device. */
+ucs_status_t uct_rc_verbs_ep_tag_connect_to_ep(uct_ep_h tl_ep,
+                                               const uct_device_addr_t *dev_addr,
+                                               const uct_ep_addr_t *ep_addr)
+{
+    ucs_status_t status;
+    uct_rc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    uct_rc_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_iface_t);
+    const uct_ib_address_t *ib_addr = (const uct_ib_address_t *)dev_addr;
+    uct_rc_verbs_ep_tm_address_t *tm_addr = (uct_rc_verbs_ep_tm_address_t*)ep_addr;
+    struct ibv_ah_attr ah_attr;
+
+    ucs_assert_always(tm_addr->super.type == UCT_RC_EP_ADDR_TYPE_TM);
+
+    uct_ib_iface_fill_ah_attr(&iface->super, ib_addr, ep->super.path_bits,
+                              &ah_attr);
+
+    /* Connect local ep QP to the one owned by device (and bound to XRQ)
+     * on the peer. */
+    status = uct_rc_iface_qp_connect(iface, ep->super.txqp.qp,
+                                     uct_ib_unpack_uint24(tm_addr->tm_qp_num),
+                                     &ah_attr, tm_addr->super.atomic_mr_id);
```

Comment:
let's remove atomic_mr_id parameter from  uct_rc_iface_qp_connect() - it's used only for debug message